Лабораторная работа 3

ZhidkovNA opened the pull request:
В task8 допущена ошибка в ответе "3"
Верный ответ "5"

![image](https://user-images.githubusercontent.com/115547518/196812574-17a39bc0-8bb4-434a-a4ac-00d22058972c.png)

ZhidkovNA committed:
Лабораторная работа 3

В task8 допущена ошибка в ответе "3"
Верный ответ "5"

andreiivanov-wa commented on Лабоаторная работа 3/task6.py:
```diff
@@ -0,0 +1,9 @@
+list_numbers = [2, -93, -2, 8, -36, -44, -1, -85, -14, 90, -22, -90, -100, -8, 38, -92, -45, 67, 53, 25]
+
```
тут под капотом стандартных функций два прохода по списку, один для поиска максимального элемента, другой для поиска индекса. Если список состоит из 1 000 000 элементов, то будет 2 000 000 "операций". Подразумевалось, что будет реализовано за один проход:
max_index = 0
max_value = list_numbers[max_index]

for i, value in enumerate(list_numbers):
····if value > max_value:
········max_value = value
········max_index = i

andreiivanov-wa replied:
но если используете стандартные функции, то комментарии ниже

ZhidkovNA replied:
Исправил код с учетом прохода в 1 раз и использованием функции (счетчика) enumerate 
https://github.com/ZhidkovNA/DD-PY1/pull/2/commits/4d395f29cf97650738dfc78124011b8ef87fd030

andreiivanov-wa commented on Лабоаторная работа 3/task6.py:
```diff
@@ -0,0 +1,9 @@
+# TODO Оформить решение
+max_num = max(list_numbers)
+max_num_index = list_numbers.index(max_num)
+last_num_index = list_numbers.index(list_numbers[-1])
```
достаточно last_num_index = -1

andreiivanov-wa commented on Лабоаторная работа 3/task8.py:
```diff
@@ -0,0 +1,22 @@
+month = 0  # количество месяцев, которое можно прожить
+
+while money_capital > 0:
+    if month == 0:
```
а чем строчки 10-12 отличаются от 14-16?

andreiivanov-wa replied:
но если подразумевать, что для условия if month == 0 должен выполняться другой код, то возникнет следующий момент. Условие month == 0 выполнится один единственный раз в самом начале, а проверяться будет каждый раз при новой итерации цикла. А зачем его проверять, если заранее известно, что оно не выполнится никогда кроме как в первом проходе. Циклы нужны для повторяющихся операций, а единичные случаи нужно вынести за пределы цикла. Например, вместо
while money_capital > 0:
····if month == 0:
········foo1()
····else:
········foo2()

следует написать
foo1()
while money_capital > 0:
····foo2()

andreiivanov-wa commented on Лабоаторная работа 3/task9.py:
```diff
@@ -0,0 +1,26 @@
+
+while months >= 1:
+    if months > 1:
+        spend = spend + spend * 0.03
```
магические константы использовать нельзя, следует использовать переменные
spend += spend * increase

andreiivanov-wa commented on Лабоаторная работа 3/task9.py:
```diff
@@ -0,0 +1,26 @@
+        #print(months)
+
+    else:
+        need_money = need_money + 6000 - salary
```
то же самое:
need_money += spend - salary

ZhidkovNA replied:
Исправил ошибки:
- Убрал цикл if - else
- Вынес операции выполняющиеся 1 раз за пределы цикла ( 1 месяц, когда не учитывается ежемесячный рост цен)
- Сократил код (a = a +1 эквивалентно a += 1)

ZhidkovNA replied:
Правильно ли выполнено изменение в функционале Github?

В разделе Commits появился файл с изменённым кодом (Update task8.py)
https://github.com/ZhidkovNA/DD-PY1/pull/2/commits/82c36525cf8c3b27572f88cfd45cb5eb69d1d553

andreiivanov-wa commented on Лабоаторная работа 3/task9.py:
```diff
@@ -0,0 +1,26 @@
+need_money = 0  # количество денег, чтобы прожить 10 месяцев
+
+while months >= 1:
+    if months > 1:
```
опять единичиный случай, но уже под else, его надо вынести за пределы цикла, в цикле он заведомо не будет отрабатывать никогда, кроме случая month == 1

ZhidkovNA replied:
Для task9.py проделал аналогичные изменения, что и для task8. py

- Убрал цикл if - else
- Вынес операции выполняющиеся 1 раз за пределы цикла ( 1 месяц, когда не учитывается ежемесячный рост цен)
- Сократил код (a = a +1 эквивалентно a += 1)

https://github.com/ZhidkovNA/DD-PY1/pull/2/commits/a6637d50fb72632c16f6a95f93354925f46d6b8f

ZhidkovNA committed:
Update task8.py

Исправил ошибки:
- Убрал цикл if - else
- Вынес операции выполняющиеся 1 раз за пределы цикла ( 1 месяц, когда не учитывается ежемесячный рост цен)
- Сократил код (a = a +1 эквивалентно a += 1)

ZhidkovNA committed:
Update task9.py

Устранены замечания (аналогично task8.py)

ZhidkovNA committed:
Update task6.py

Исправил код с учетом прохода в 1 раз и использованием функции (счетчика) enumerate
